polaris/polaris/sep10auth/views.py

```diff
@@ -29,150 +35,179 @@
+        of the signers on the challenge are signers for the account and the medium
+        threshold on the account is met by those signers.
+
+        If the source account does not exist, verify that the keypair used as the
+        source for the challenge transaction has signed the challenge. This is
+        sufficient because newly created accounts have their own keypair as signer
+        with a weight greater than the default thresholds.
+        """
+        server_key = settings.SIGNING_KEY
+        net = settings.STELLAR_NETWORK_PASSPHRASE
+
+        logger.info("Validating challenge transaction")
+        try:
+            tx_envelope, account_id = read_challenge_transaction(
+                envelope_xdr, server_key, net
+            )
+        except InvalidSep10ChallengeError as e:
+            err_msg = f"Error while validating challenge: {str(e)}"
+            logger.error(err_msg)
+            raise ValueError(err_msg)
+
+        try:
+            account = settings.HORIZON_SERVER.load_account(account_id)
+        except NotFoundError:
+            logger.warning(
+                "Account does not exist, using client's master key to verify"
+            )
+            try:
+                verify_challenge_transaction_signed_by_client(
+                    envelope_xdr, server_key, net
+                )
+            except InvalidSep10ChallengeError as e:
+                logger.info(
+                    f"Missing or invalid signature(s) for {account_id}: {str(e)})"
+                )
+                raise ValueError(str(e))
+            else:
+                logger.info("Challenge verified using client's master key")
+                return
+
+        signers = account.load_ed25519_public_key_signers()
+        threshold = account.thresholds.med_threshold
+        try:
+            signers_found = verify_challenge_transaction_threshold(
+                envelope_xdr, server_key, net, threshold, signers
+            )
+        except InvalidSep10ChallengeError as e:
+            logger.info(str(e))
+            raise ValueError(str(e))
+
+        logger.info(f"Challenge verified using account signers: {signers_found}")
+
+    @staticmethod
+    def _generate_jwt(request: Request, envelope_xdr: str) -> str:
+        """
+        Generates the JSON web token from the challenge transaction XDR.
+
+        See: https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#token
+        """
+        issued_at = time.time()
+        transaction_envelope, source_account = read_challenge_transaction(
+            envelope_xdr, settings.SIGNING_KEY, settings.STELLAR_NETWORK_PASSPHRASE
+        )
+        logger.info(
+            f"Challenge verified, generating SEP-10 token for account {source_account}"
+        )
+        hash_hex = binascii.hexlify(transaction_envelope.hash()).decode()
+        jwt_dict = {
+            "iss": request.build_absolute_uri("/auth"),
```

Comment:
I think this line allows injection into the JWT. If an anchor hosts their server in anyway that allows arbitrary hostnames in the requests, e.g. if they host it on a static IP without hostname filtering which is possible, the `iss` field can be made to contain any hostname a caller wants.

I'd lean towards this being the server signing key address, e.g. https://github.com/stellar/go/blob/4268b290621430b89c3ed687e0d83df3800c227a/exp/services/webauth/internal/serve/token.go#L87, if we don't have a good way to get some other definitive exact URI for the issuer.

Reply:
I'm not convinced that using the server signing key is any better than the current value for `iss`; both can be faked. The only real protection against someone attempting to side-step SEP-10 by using a fake JWT (a JWT not created by the anchor server) is the server's secret used to encode the data. The SEP also states that `iss` should be a URI. What do you think?

Reply:
Ah you're right, the SEP does say to use a URI. I was thinking about the RFC7519 Section 4.1.1 where technically it doesn't have to be a URI, but I forgot we specify that in the SEP. I'll change the `webauth` implementation to make it a URI too.

Can we get the hostname from the settings or somewhere else instead of the incoming request? I still think we should avoid allowing any data from the request making its way into the JWT.

Reply:
FYI I updated the Go implementation here to align with ☝️: stellar/go#2214.

Reply:
Yea, I can add an `AUTH_SERVER` setting since the SEP-10 authentication could be hosted on a different URI than the anchor server.

Reply:
Actually, if they're using Polaris there isn't a reason you'd want to have a separate server handling authentication. I'll use a `HOST_URL` instead.